### PR TITLE
Introducing webhooks (WIP)

### PR DESCRIPTION
**Pull request by waptik**
## Intro
This PR aims to introduce a new feature inside **FrameTrain** called **Webhooks**.
Webhooks will allow **Users** receive data on frame **Viewers** interactions from FrameTrain directly to their server through the webhook url they provide per event based on the events and kind of interactions configuration defined by the frame **Developer**.

See https://github.com/FTCHD/frametrain/pull/74#issuecomment-2225548003 for more info

### CODE REVIEW

**Comment by waptik** [comment]
Here are the latest updates as of 12th July 2024, 12:28PM

- Devs (template makers), can add a list of events in the events array inside the default export of their templates `index.ts` file
image

- For each event they want to send webhook data to a user's server, they(Devs) have to define the `webhook` object inside a `handler` (one event per handler) of their choosing in following format:
 ```ts
return {
        ...,
        webhook: {
            event: string, // event to trigger the webhook
            data: Record<string, any>, // data to send to the webhook
        },
    }

 ```
image

- If the Dev has set some events inside `index.ts`, the User will see a button a top bar labeled **Webhooks**
image

- Upon clicking on the webhooks button, the user will see a list of events and toggle beside each event.
image

- Switching on a toggle will render a url input field to enter the the webhook url where data for this event should be sent.
image

- If the User attached a url to specific event in the popover, they will receive a webhook event with some data based on the structure set by the dev
image


- No data will be sent if the input field is empty even if the toggle is switched on, or the event's toggle has been switched off

### User Demo
https://www.loom.com/share/4a7a8a29b4d441468dbcd4fdfe0bea9b

File: components/editor/WebhookEventOptions.tsx [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    DropdownMenuTrigger,
+} from '@/components/shadcn/DropdownMenu'
+import { Switch } from '@/components/shadcn/Switch'
+import { useState } from 'react'
+import { Label } from '../shadcn/InputLabel'
+import { Input } from '../shadcn/Input'
+import { useDebouncedCallback } from 'use-debounce'
+import { updateFrameWebhooks } from '@/lib/frame'
+
+export default function WebhookEventOptions(props: {
```
Comment (FTCHD) [inline components/editor/WebhookEventOptions.tsx:19]: this should be 
```ts
{
    events,
    webhooks,
    id,
    setUpdating,
}: {
    events: string[]
    webhooks: { [key: string]: string }
    id: string
    setUpdating: (updating: boolean) => void
}
```

File: components/editor/WebhookEventOptions.tsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                {/* <DropdownMenuItem>
+                    <span className="w-full">FID</span>
+                    <Input
+                        defaultValue={mockOptions.fid}
+                        onChange={async (e) => {
+                            const fid = Number.parseInt(e.target.value)
+                            setMockOptions((prev) => ({ ...prev, fid }))
+                        }}
+                    />
+                </DropdownMenuItem> */}
```
Comment (FTCHD) [inline components/editor/WebhookEventOptions.tsx:107]: ?

File: lib/farcaster.d.ts [hunk 3]
```diff
@@ -30,4 +30,8 @@ export interface BuildFrameData {
     component?: ReactElement
     image?: string
     handler?: string
+    webhook?: {
```
Comment (FTCHD) [inline lib/farcaster.d.ts:33]: should be `webhooks`, meaning an array. a handler can trigger multiple events.

File: lib/serve.ts [hunk 4]
```diff
@@ -34,6 +38,10 @@ export async function buildFramePage({
 
     let imageData
 
+    if (webhook) {
```
Comment (FTCHD) [inline lib/serve.ts:41]: triggering events happens in the `/f/[handler]...` route, and only there.

File: lib/serve.ts [hunk 5]
```diff
@@ -270,3 +287,82 @@ export async function validatePayload(
...
+    // get the webhook url for the event
+    const webhookUrl = frame.webhooks[event] as string | undefined
+
+    if (!webhookUrl) {
+        return
+    }
+
+    const id = crypto.randomUUID()
+
+    for (let i = 0; i <= 3; i++) {
```
Comment (FTCHD) [inline lib/serve.ts:328]: remove retry logic and `id`

File: lib/serve.ts [hunk 6]
```diff
@@ -270,3 +287,82 @@ export async function validatePayload(
...
+ * It sends a POST request to the webhook url.
+ * There's a limit of 3 tries at an interval of 5 seconds each
+ *
+ * @param frameId string - the frame id
+ * @param event string - an object of the event name and external url
+ * @param data object - data to be sent with the event
+ * @returns Promise<void>
+ */
+async function triggerEvent(frameId: string, event: string, data: Record<string, unknown>) {
+    const frame = await client
```
Comment (FTCHD) [inline lib/serve.ts:308]: create a `getFrameWebhooks` function in `lib/frame.ts`. 

remove the triggerEvent function completely. in the `/f/[handler]` route you check if we have `webhooks` in the buildParameters, if so we `getFrameWebhooks`, iterate over all buildParameters.webhooks and send a request if the user has an endpoint setup for the event.

`webhooks` is not used in the `buildFramePage` function, you can remove it from the parameters entirely.

File: components/editor/WebhookEventOptions.tsx [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+                    </div>
+                </DropdownMenuLabel>
+
+                <DropdownMenuSeparator />
+
+                <DropdownMenuGroup>
+                    {props.events.map((event) => (
+                        <DropdownMenuItem className="flex-col gap-4" key={event}>
+                            <div className="flex items-start w-full">
+                                <span className="w-full">{event}</span>
```
Comment (FTCHD) [inline components/editor/WebhookEventOptions.tsx:65]: `${event[0].toUpperCase()}${event.slice(1)}`

Comment (waptik) [reply components/editor/WebhookEventOptions.tsx]: Correction: `{event[0].toUpperCase()}{event.slice(1)}`

File: templates/cal/index.ts [hunk 8]
```diff
@@ -54,8 +54,7 @@ export default {
     initialConfig: {
         events: [],
         bio: [],
-        gatingOptions: {
-            karmaGating: false,
+        g: {
```
Comment (FTCHD) [inline templates/cal/index.ts:57]: if this change is valid, then please open a separate PR

File: app/(app)/f/[frameId]/[handler]/[[...query]]/route.ts [hunk 9]
```diff
@@ -97,15 +97,50 @@ export async function POST(
...
     if (storageData) {
         await updateFrameStorage(frame.id, storageData)
         console.log('Updated frame storage')
     }
     await updateFrameCalls(frame.id, frame.currentMonthCalls + 1)
 
+    // Allow for important logic to get processed and don't wait for webhooks to finish
+    if (buildParameters.webhooks?.length) {
+        try {
+            const webhookUrls = await getFrameWebhooks(frame.id)
```
Comment (FTCHD) [inline app/(app)/f/[frameId]/[handler]/[[...query]]/route.ts:112]: on a second thought this is necessary, so remove the function call and the function itself from `lib/frame`. the frame object is already a few lines above, so `webhookUrls = frame.webhooks`.

File: components/FrameEditor.tsx [hunk 10]
```diff
@@ -9,7 +9,7 @@ import {
     updateFrameName,
 } from '@/lib/frame'
 import { previewParametersAtom } from '@/lib/store'
-import type templates from '@/templates'
+import templates from '@/templates'
```
Comment (FTCHD) [inline components/FrameEditor.tsx:12]: this was correct before, why does this PR change it?

Comment (waptik) [reply components/FrameEditor.tsx:12]: Sorry, it happened during merge-conflicts resolution this morning.

File: templates/presentation/index.ts [hunk 11]
```diff
@@ -36,49 +36,45 @@ export interface Slide {
...
-		weight: '400',
-	},
-	buttons: [
-		{
-			type: 'navigate',
-			label: 'Navigate',
-			target: '1',
-		},
-	],
+const DEFAULT_SLIDE: Slide = {
```
Comment (waptik) [inline templates/presentation/index.ts:43]: The cahnges in this block are as a result of using biome format whilst i was formatting directories i've touched

**Pull request merged**